microprofile-jwt/src/test/java/org/jboss/eap/qe/microprofile/jwt/security/publickeylocation/PublicKeyPropertyTestCase.java

```diff
@@ -0,0 +1,128 @@
+import java.net.URL;
+import java.util.concurrent.TimeUnit;
+import java.util.regex.Pattern;
+
+import static org.hamcrest.CoreMatchers.equalTo;
+
+/**
+ * Set of tests verifying functionality of {@code mp.jwt.verify.publickey} property.
+ */
+@RunWith(Arquillian.class)
+public class PublicKeyPropertyTestCase {
+
+    private static final String DEPLOYMENT_WITH_VALID_KEY = "valid-key-deployment";
+    private static final String DEPLOYMENT_WITH_INVALID_KEY = "invalid-key-deployment";
+
+    private static RsaKeyTool keyTool;
+
+    @BeforeClass
+    public static void beforeClass() throws URISyntaxException {
+        final URL privateKeyUrl = BasicCdiTest.class.getClassLoader().getResource("pki/key.private.pkcs8.pem");
+        if (privateKeyUrl == null) {
+            throw new IllegalStateException("Private key wasn't found in resources!");
+        }
+        keyTool = RsaKeyTool.newKeyTool(privateKeyUrl.toURI());
+    }
+
+    /**
+     * A deployment with valid key. This means, that it is correct key which can be used to verify the JWT signature.
+     * @return a deployment
+     */
+    @Deployment(name = DEPLOYMENT_WITH_VALID_KEY)
+    public static WebArchive createValidKeyDeployment() {
+        return ShrinkWrap.create(WebArchive.class)
+                .addClass(JaxRsBasicEndpoint.class)
+                .addClass(JaxRsTestApplication.class)
+                .addAsManifestResource(BasicCdiTest.class.getClassLoader().getResource("mp-config-pk-valid.properties"),
+                        "microprofile-config.properties");
+    }
+
+    /**
+     * A deployment with invalid key. The verification fails when this key is used.
+     * @return a deployment
+     */
+    @Deployment(name = DEPLOYMENT_WITH_INVALID_KEY)
+    public static WebArchive createInvalidKeyDeployment() {
+        return ShrinkWrap.create(WebArchive.class)
+                .addClass(JaxRsBasicEndpoint.class)
+                .addClass(JaxRsTestApplication.class)
+                .addAsManifestResource(BasicCdiTest.class.getClassLoader().getResource("mp-config-pk-invalid.properties"),
+                        "microprofile-config.properties");
+    }
+
+    /**
+     * @tpTestDetails A request with proper JWT is sent on server which has configured bad public key. The server fails
+     * to verify the JWT signature and won't authorize the client. Test specification compatibility.
+     * @tpPassCrit "Unauthorized" message is shown to user and warning is present in log specifying the cause of fail.
+     * @tpSince EAP 7.3.0.CD19
+     */
+    @Test
+    @RunAsClient
+    @OperateOnDeployment(DEPLOYMENT_WITH_INVALID_KEY)
+    public void testJwkVerificationFailsWithInvalidKey(@ArquillianResource URL url) throws InterruptedException {
+        final JsonWebToken token = new JwtHelper(keyTool, "issuer").generateProperSignedJwt();
+
+        final LogListener listener = new LogListener(Pattern.compile(".*WARN.*Token is invalid: JWT rejected due to invalid signature.*"));
```

Comment:
TBH I don't like log parsing

And you are not even searching for message ID here

You should use Java API or Byteman for these checks

Reply:
I do this because customer will do exactly same thing - something will go wrong, they will check logs.

Good point - thank you! This is actually propagated message from smallrye and we probably don't want that...

Which Java API do you have on mind?

Reply:
it's more general comment, Java API - to inject stuff and do checks on it
e.g. https://github.com/eclipse/microprofile-metrics/blob/master/tck/api/src/main/java/org/eclipse/microprofile/metrics/tck/metrics/CountedMethodTagBeanTest.java#L70 way 

In this case Byteman rule against logging is probably the right way

Reply:
Asked on https://wildfly.zulipchat.com/#narrow/stream/210028-microprofile/topic/logging.20-.20programatic.20checks/near/181918661

Reply:
James Perkins 2:32 PM   
There is some documentation here https://docs.wildfly.org/18/Admin_Guide.html#logging-filters. As of WildFly 18 you can use custom filters which is probably what you want.

Reply:
Thank you! This will need some hacking*, but it might actually solve the problem...

* filter class must reside in WildFly module and we need mechanism to communicate with filter to check if filter intercepted desired message.